### PR TITLE
Optimized `SCC` Implementation and Removed `TensorFlow` Dependencies

### PR DESCRIPTION
**Pull request by Starlitnightly**
When installing `spateo` on macOS, I encountered version incompatibility errors. For instance, the `vtk` package requires a minimum Python version of 3.9. Additionally, installing both `torch` and `tensorflow` simultaneously can lead to package conflicts. After thoroughly reviewing the implementation of the relevant TensorFlow code, I rewrote all TensorFlow-related code using PyTorch and removed all TensorFlow-related dependencies from `requirements.txt`.

### Updates:
- Re-Implemented `NLPCA` using PyTorch instead of TensorFlow.
- Re-Implemented `weighted_binary_crossentropy` using PyTorch instead of TensorFlow.
- Re-implemented `calculate_leiden_partition`and added `logger.info`.
- Optimized label assignment by majority voting in the neighborhood using `optimize_cluster`.

### Notes:
After carefully comparing the clustering effects of Leiden in `scanpy` and `spateo`, I found that the enhanced effect is due to `dynamo.tl.neighbors` compared to `scanpy.pp.neighbors`. The exact reason for this enhancement still needs further investigation.

Additionally, it is important to note that `scc` does not achieve state-of-the-art (SOTA) performance and does not yield better results on the gold standard dataset of human cortical neurons.

Since `scc` is an adjacency matrix that directly combines spatial neighborhoods and PCA neighborhoods, mclust is not applicable and was not included in this PR. Perhaps I could introduce STAGATE into the `spateo` framework and use it in place of PCA. However, this would introduce a new dependency, [`pyg`](https://pytorch-geometric.readthedocs.io/), which might complicate updating the existing `requirements.txt` and incur additional installation costs for users.

image

image

image

### CODE REVIEW

**Comment by Starlitnightly** [comment]
More info about `optimize_cluster `

image

**Comment by Sichao25** [comment]
Those updates are awesome; thanks for your contribution! I noticed you've specified matplotlib<=3.6.2 and pandas<=1.5.3 in your changes. Is there a specific reason for keeping these versions? We're currently updating these dependencies, and I also feel they may be causing some dependency issues in the CI.

**Comment by Starlitnightly** [comment]
Thank you for your feedback! I’m glad you found the updates helpful.

Regarding the specific versions of `matplotlib` and `pandas`, there are a couple of reasons for those choices:

1. **Pandas**: Version 1.5.3 is the last release before Pandas 2.0, which replaced the internal engine with Polars. This update introduced significant changes, including the renaming of many functions and alterations in formatting. Consequently, many packages have limited compatibility to Pandas 1.5.3 to avoid these disruptions.

2. **Matplotlib**: For `matplotlib`, the version restriction to 3.6.2 or 3.6.3 is due to changes in the parameters of some plotting functions in higher versions, such as `cmap`. To maintain consistency and avoid potential issues, we opted to restrict it to 3.6.2. Without this restriction, we would need to implement version checks before calling plotting functions to ensure compatibility.

However, I understand the importance of keeping our dependencies up-to-date. If the current versions are causing CI issues, I’m more than willing to help investigate and test with the latest versions to ensure everything works smoothly.

**Comment by Starlitnightly** [comment]
@Sichao25 But the latest version of anndata seems to have solved the support bug for pandas versions greater than 2. This requires more testing to ensure compatibility

scverse/anndata#948

**Comment by Sichao25** [comment]
Thanks for sharing your concerns. We are updating important dependencies like `NumPy`, `pandas`, and `matplotlib` to ensure compatibility with other tools. The current plan includes support for pandas >=1.5.3 and matplotlib>=3.6.2. These updates will make it easier for you to pass the CI and test your changes. Feel free to join the discussion [here](https://github.com/aristoteleo/spateo-release/issues/244) to share your ideas or report any dependency issues.

About scc, it will be nice to have STAGATE introduced as a new option. How about setting it as an optional dependency that users can install themselves? Specifically, we won't add new dependency directly to `requirements.txt`, but we will check for its installation before using it by some code like:
```
try:
    import STAGATE
except ImportError:
```

File: spateo/tools/cluster/find_clusters.py [hunk 1]
```diff
@@ -233,5 +233,43 @@ def scc(
...
+def optimize_cluster(adata: anndata.AnnData, radius: int = 50, key: str = "label") -> list:
+    """
+    Optimize the label by majority voting in the neighborhood.
+
+    Args:
+        adata: an Anndata object, after normalization.
+        radius: the radius of the neighborhood.
+        key: the key in `.obs` that corresponds to the cluster labels.
+    """
+    import ot
```
Comment (Xiaojieqiu) [inline spateo/tools/cluster/find_clusters.py:251]: are you using POT? POT has some dependency issues and we are planning to remove it. I think we can use SciPy instead of ot for distance calculations.

File: spateo/tools/cluster/leiden.py [hunk 2]
```diff
@@ -81,23 +81,43 @@ def calculate_leiden_partition(
...
-            for i, j in zip(*adj.nonzero()):
-                G.add_edge(i, j)
+            adj = scipy.sparse.csr_matrix(adj)
+        sources, targets = adj.nonzero()
+        weights = adj[sources, targets]
+        if isinstance(weights, np.matrix):
+            weights = weights.A1
+        G = igraph.Graph(directed=None)
+        G.add_vertices(adj.shape[0])  # this adds adjacency.shape[0] vertices
+        G.add_edges(list(zip(sources, targets)))
```
Comment (Xiaojieqiu) [inline spateo/tools/cluster/leiden.py:103]: is this function  G.add_edges(list(zip(sources, targets))) fast? is it as fast as adj_to_igraph(adj.tolist())?

**Review by Xiaojieqiu — COMMENTED** [review]
Hi @Starlitnightly thanks for this pull request. Your ideas on switching from tensorflow to pytorch is a great one. I also like your optimize_cluster function which can definitely be used to smooth the cluster layer on the space

**Comment by Xiaojieqiu** [comment]
@Starlitnightly regarding your comments on scc, we found scc works well on several dataset given its simple formulation. (1) but use STAGATE to replace pca seem like a good idea. (2) are you saying spateo's leiden results are worse than scanpy?

File: spateo/tdr/models/models_backbone/backbone_methods.py [hunk 3]
```diff
@@ -61,17 +63,23 @@ def fit(self, data: np.ndarray, epochs: int = 500, nodes: int = 25, lr: float =
...
-        bname = model.layers[2].name  # bottle-neck layer name
+        # create model
+        self.model = self.create_model(num_dim, nodes=nodes)
+        optimizer = optim.Adam(self.model.parameters(), lr=lr)
+
+        data_tensor = torch.tensor(data, dtype=torch.float32)
 
-        # The itermediate model gets the output of the bottleneck layer,
-        # which acts as the projection layer.
-        self.intermediate_layer_model = Model(inputs=model.input, outputs=model.get_layer(bname).output)
```
[the remark this answers is not in the document]
Comment (Starlitnightly) [reply spateo/tdr/models/models_backbone/backbone_methods.py:70]: I implemented `intermediate_layer_model` inside the AutoEncoder later on, which should really be removed in self, I need to test it a bit further

File: spateo/tools/cluster/leiden.py (hunk 2 above)
Comment (Starlitnightly) [reply spateo/tools/cluster/leiden.py:103]: Yes, I directly utilized the properties of sparse matrices here, replacing the original for loop iteration for adding nodes, which reduced the time complexity. This not only improved the algorithm's efficiency but also reduced memory usage.

Comment (Starlitnightly) [reply spateo/tools/cluster/leiden.py:103]: You're right, `adj_to_igraph(adj.tolist())` would indeed be faster, but both are faster than a for loop to add nodes

**Comment by Starlitnightly** [comment]
> @Starlitnightly regarding your comments on scc, we found scc works well on several dataset given its simple formulation. (1) but use STAGATE to replace pca seem like a good idea. (2) are you saying spateo's leiden results are worse than scanpy?

I compared the implementation of leiden in spateo and scanpy, in fact both are the same, so the result is the same, it is the **`neighbors`** that cause the difference in leiden. spateo's results are better compared to scanpy's

File: spateo/tools/cluster/find_clusters.py (hunk 1 above)
Comment (Starlitnightly) [reply spateo/tools/cluster/find_clusters.py]: I replaced the `POT` with `scipy` in the new Pull Request, although I lost some of the runtime efficiency.

File: spateo/tdr/models/models_backbone/backbone_methods.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Starlitnightly) [reply spateo/tdr/models/models_backbone/backbone_methods.py:70]: I have tested this function on simulated data, and it runs well. However, I haven't found specific use cases for this function. Could someone provide specific scenarios where this function is used so that I can further optimize it?

**Comment by Starlitnightly** [comment]
> Thanks for sharing your concerns. We are updating important dependencies like `NumPy`, `pandas`, and `matplotlib` to ensure compatibility with other tools. The current plan includes support for pandas >=1.5.3 and matplotlib>=3.6.2. These updates will make it easier for you to pass the CI and test your changes. Feel free to join the discussion [here](https://github.com/aristoteleo/spateo-release/issues/244) to share your ideas or report any dependency issues.
> 
> About scc, it will be nice to have STAGATE introduced as a new option. How about setting it as an optional dependency that users can install themselves? Specifically, we won't add new dependency directly to `requirements.txt`, but we will check for its installation before using it by some code like:
> 
> ```
> try:
>     import pyg
> except ImportError:
> ```

I'm going to try to implement STAGATE in spateo in a future PR. Since the author doesn't give a usable package directly in `pypi`, a lot of files and code may need to be added!

File: spateo/tools/cluster/leiden.py (hunk 2 above)
Comment (Xiaojieqiu) [reply spateo/tools/cluster/leiden.py:103]: then maybe we can use igraph.Graph.Weighted_Adjacency(adj) to replace all these lines of code:

```
        if isinstance(weights, np.matrix):
            weights = weights.A1
        G = igraph.Graph(directed=None)
        G.add_vertices(adj.shape[0])  # this adds adjacency.shape[0] vertices
        G.add_edges(list(zip(sources, targets)))
```

**Comment by Starlitnightly** [comment]
Maybe it need to be added more method for cluster SOTA in this commit.

- [ ] STAGATE: https://github.com/QIFEIDKN/STAGATE_pyG
- [ ] CAST: https://github.com/wanglab-broad/CAST/blob/main/setup.py

**Comment by Starlitnightly** [comment]
The latest tutorial can be viewed in the pull request of spateo-tutorial: https://github.com/aristoteleo/spateo-tutorials/blob/bf32d5739f380948e76cbe07c99e3e8d6d1e3627/5_cluster_digitization/1_bin_scc.ipynb

It should be noted that CAST and STAGATE do not perform well on the current dataset, but they outperform SCC on other datasets. Additionally, in the new tutorial, I have modified the method of spatial domain annotation to use dictionary-based annotation instead of sequential annotation.

**Comment by Xiaojieqiu** [comment]
excellent work. I am going to merge the pull request now.
